File: OctavePSSProgram/Smoother.m
# The Smoother class is responsible for taking salted data and making it more
# understandable.
#
# Author: Rachel Hussmann
classdef Smoother

  # This section is like the body of a Java class.
  # This section is stating that the access to the method is public and the
  # class has one method.
  methods (Access = public)

    # The smoothData method accepst the salted y values and averages them with
    # a window of values around it to smooth the line and make the data more
    # understandable.
    #
    # Parameters: smooth - The Smoother object, yValues - The salted y values,
    # windowValue - The number of data points to be averaged to smooth the data,
    # runs - The number of times the smoothData method should be run
    #
    # Returns: smoothY - The smoothed y values
    function smoothY = smoothData(smooth, yValues, windowValue, runs)

      # Create the tempYValues list so it can hold all of the new averaged
      # values of y
      tempYValues = [];

      # Runs the program number of times the user wanted
      for c = 1 : runs

        # Iterate through the list
        for i = 1 : length(yValues)

          #Create the sum and averageCount variables to calculated the average
          # at the end
          totalSum = 0.0;
          averageCount = 0;

          # Iterate through the values to the left of the current index
          for j = 1 : windowValue

           # Check to see if the value would be out of bounds
           if i-j >= 1

              # If the value is not out of bounds, add the value to the sum and
              # add one to the average count
              totalSum += yValues(i-j);
              averageCount += 1;
            endif
          endfor

          # Add the current value to the sum and the increment the average count
          totalSum += yValues(i);
          averageCount += 1;

          # Iterate through the values to the right of the current index
          for j = 1 : windowValue

            # Check to see if the value would be out of bounds
            if i+j <= length(yValues)

              # If the value is not out of bounds, add the value to the sum and
              # add one to the average count
              totalSum += yValues(i+j);
              averageCount += 1;
            endif

          endfor

          # Add the average to the end of the list of tempYValues
          tempYValues(end+1) = totalSum/averageCount;

        endfor

        # Set yValues equal to tempYValues and empty tempYValues
        yValues = tempYValues;
        tempYValues = [];

      endfor

      # When the whole program is done, return the smoothed yValues
      smoothY = yValues;

    endfunction

  endmethods

endclassdef

File: PSS/src/Smoother.java
import java.util.ArrayList;

/**
 * The Smoother class is responsible for taking salted data and making it more understandable.
 * @author Rachel Hussmann
 */
public class Smoother {

    private ArrayList<Integer> xValues;
    private ArrayList<Double> yValues;

    /**
     * Additional constructor for the Smoother class
     * @param yValues The salted y values from the function
     */
    public Smoother(ArrayList<Double> yValues){
        this.xValues = new ArrayList<>();
        this.yValues = yValues;
    }

    /**
     * The smoothData method takes in salted values and finds the average of a set number of values to smooth the data.
     * @param windowValue The number of data points to be averaged to smooth the data
     */
    public void smoothData(int windowValue){

        //Create the ArrayLists to save the x and y values
        ArrayList<Integer> tempXValues = new ArrayList<>();
        ArrayList<Double> tempYValues = new ArrayList<>();

        //Initialize the xCount value to keep track of the x values
        int xCount = 1;

        //Outer for loop iterates through the salted dataset
        for(int i = 0; i < yValues.size(); i++){

            //Initialize the sum and average count needed to create the averaged values
            double sum = 0.0;
            int averageCount = 0;

            //First for loop adds up the values to the left of the current index and the current index
            for(int j = i - windowValue; j <= i; j++){

                //Check to see if the value is outside the range of the list
                if(j >= 0){

                    //If the value is within the bounds of the list, add it to the sum and increment the average count
                    sum += yValues.get(j);
                    averageCount++;
                }
            }

            //Second for loop adds up the right side values (excluding the current index because that was handled in
            //the first for loop)
            for(int j = i + 1; j <= i + windowValue; j++){

                //Check to see if the value is outside the range of the list
                if(j < yValues.size()){

                    ////If the value is within the bounds of the list, add it to the sum and increment the average count
                    sum += yValues.get(j);
                    averageCount++;
                }
            }

            //Add the current count of x to the tempXValues and increment the xCount variables by 1
            tempXValues.add(xCount);
            xCount++;

            //Add the average of the window to the tempYValues arraylist
            tempYValues.add(sum/averageCount);
        }

        //Clear out the xValues and yValues arraylists to copy the new values into them
        xValues.clear();
        xValues.addAll(tempXValues);
        yValues.clear();
        yValues.addAll(tempYValues);
    }

    /**
     * The saveSmoothedFunction method saves the x values and the smoothed y values into a .CSV file
     */
    public void saveSmoothedFunction(){
        //Create the exporter object
        Exporter exporter = new Exporter();

        //Call the createFile method to save the values in a csv file
        exporter.createFile(xValues, yValues, "SmoothedLogFunction.csv");
    }

    /**
     * The runSmoother method is responsible for running the smoothData method the number of times the user would like
     * and saving the results of the smoothing.
     * @param windowValue The number of values to left and to the right of each value to average
     * @param numberOfRuns The number of times to run the smoothData method
     */
    public void runSmoother(int windowValue, int numberOfRuns){

        //Runs the smoothData method the user requested number of times
        for(int i = 0; i < numberOfRuns; i++){

            //Call the smoothData method
            smoothData(windowValue);
        }

        //Save the data that has been smoothed
        saveSmoothedFunction();
    }
}
